### ISSUE TITLE
checkForUpdateAsync() crashes app in Android (expo-updates 0.18.16)

### ISSUE THREAD

**Issue by leons1767** [expo/expo#24849]
### Minimal reproducible example

https://snack.expo.dev/@ad9889/expo-checkforupdateasync-crash

### Summary

The newly built Android app will crash when Updates.checkForAsync() is ran without any expo update published. The way to bypass is to publish an expo update for Android even without any code change. Once the update is published, the app may still crash for one time, before it is able to install the update. After that, it will launch normally without the crash.

At first, Sentry wasn't able to pinpoint the crash to Updates.checkForAsync(). It pointed to a component that was running when the app crashed. Once the first update was published, the app crashed one more time before it was able to get the update. This time, Sentry gave the follow.
```
ArrayIndexOutOfBoundsException: length=16; index=268435453
expo.modules.updates.UpdatesUtils in bytesToHex at line 267
expo.modules.updates.loader.Loader$downloadAllAssets$1 in onFailure at line 250
expo.modules.updates.loader.FileDownloader$downloadAsset$1 in onFailure at line 588
expo.modules.updates.loader.FileDownloader$downloadFileAndVerifyHashAndWriteToPath$1 in onResponse at line 95
expo.modules.updates.loader.FileDownloader$downloadData$1 in onResponse at line 622
okhttp3.internal.connection.RealCall$AsyncCall in run at line 519
java.util.concurrent.ThreadPoolExecutor in runWorker at line 1167
java.util.concurrent.ThreadPoolExecutor$Worker in run at line 641
java.lang.Thread in run at line 923
``` 

### Environment

```
expo-env-info 1.0.5 environment info:
    System:
      OS: Windows 10 10.0.22621
    Binaries:
      Node: 18.16.0 - C:\Program Files\nodejs\node.EXE
      Yarn: 1.22.19 - ~\AppData\Roaming\npm\yarn.CMD
      npm: 8.19.2 - ~\AppData\Roaming\npm\npm.CMD
      Watchman: 20230416.213759.0 - D:\ReactNative\watchman-v2023.04.17.00-windows\bin\watchman.EXE
    npmPackages:
      @expo/webpack-config: ^19.0.0 => 19.0.0
      expo: ^49.0.13 => 49.0.13
      react: 18.2.0 => 18.2.0
      react-dom: 18.2.0 => 18.2.0
      react-native: ^0.72.5 => 0.72.5
      react-native-web: ~0.19.6 => 0.19.8
    Expo Workflow: bare
```

**Comment by Chrissi2812** [comment]
This also affects iOS Builds.

And thanks for that "workaround" that really saved me 👍
Hope this will get fixed soon.

**Comment by douglowder** [comment]
I am not able to reproduce any crash with the provided Snack, either with SDK 48 or 49.

**Comment by leons1767** [comment]
> I am not able to reproduce any crash with the provided Snack, either with SDK 48 or 49. If you can provide detailed steps on how to set up the code in the Snack to reproduce a crash, I'll reopen this issue.

May I know if you build the production apk and run it? Expo Go and custom development build are blocked from checking update by design, so it can't be reproduced in Snack. A production build is required to reproduce the issue.

If I catch the error in development environment, it returns the below. 

Failed to check for EAS update: You cannot check for updates in development mode. To test manual updates, make a release build with `npm run ios --configuration Release` or `npm run android --variant Release`.

**Comment by douglowder** [comment]
@leons1767 indeed I ran a release build when testing this.

**Comment by leons1767** [comment]
Thanks @douglowder. I confirm the problem is not with checkForUpdateAsync(). I'm sorry for taking up your time unnecessarily. 

I traced the problem to react-native-dotenv@3.4.8. Somehow the environment variables are not being parsed during eas build for Android, but it is done correctly during eas update, and in development build.

I just noticed I can use process.env directly with Expo SDK49. I will be taking out react-native-dotenv to save me the trouble of further troubleshooting.

Thanks again for your time.
